src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,1005 @@
+        /// <remarks>Doesn't support asking for the same hash twice before getting a response.</remarks>
+        /// <param name="headers">Collection of consecutive headers (but gaps are ok: a1=a2=a3=a4=a8=a9).</param>
+        void RequestBlocksDownload(List<ChainedHeader> headers);
+
+        /// <summary>Removes assignments for the block which has been delivered by the peer assigned to it and calls the callback.</summary>
+        /// <remarks>
+        /// This method is called for all blocks that were delivered. It is possible that block that wasn't requested
+        /// from that peer or from any peer at all is delivered, in that case the block will be ignored.
+        /// It is possible that block was reassigned from a peer who delivered it later, in that case it will be ignored from this peer.
+        /// </remarks>
+        /// <param name="blockHash">The block hash.</param>
+        /// <param name="block">The block.</param>
+        /// <param name="peerId">ID of a peer that delivered a block.</param>
+        void PushBlock(uint256 blockHash, Block block, int peerId);
+    }
+
+    /// <inheritdoc cref="IBlockPuller"/>
+    public class BlockPuller : IBlockPuller, IDisposable
+    {
+        /// <summary>Interval between checking if peers that were assigned important blocks didn't deliver the block.</summary>
+        private const int StallingLoopIntervalMs = 500;
+
+        /// <summary>The minimum empty slots percentage to start processing <see cref="downloadJobsQueue"/>.</summary>
+        private const double MinEmptySlotsPercentageToStartProcessingTheQueue = 0.1;
+
+        /// <summary>
+        /// Defines which blocks are considered to be important.
+        /// If requested block height is less than out consensus tip height plus this value then the block is considered to be important.
+        /// </summary>
+        private const int ImportantHeightMargin = 10;
+
+        /// <summary>The maximum time in seconds in which peer should deliver an assigned block.</summary>
+        /// <remarks>If peer fails to deliver in that time his assignments will be released and the peer penalized.</remarks>
+        private const int MaxSecondsToDeliverBlock = 30; //TODO change to target spacing / 3
+
+        /// <summary>This affects quality score only. If the peer is too fast don't give him all the assignments in the world when not in IBD.</summary>
+        private const int PeerSpeedLimitWhenNotInIbdBytesPerSec = 1024 * 1024;
+        
+        /// <param name="blockHash">Hash of the delivered block.</param>
+        /// <param name="block">The block.</param>
+        public delegate void OnBlockDownloadedCallback(uint256 blockHash, Block block);
+
+        /// <summary>Callback which is called when puller received a block which it was asked for.</summary>
+        /// <remarks>Provided by the component that creates the block puller.</remarks>
+        private readonly OnBlockDownloadedCallback OnDownloadedCallback;
```

Comment:
my resharper complains this should be camel case so `onDownloadedCallback`

Reply:
Fixed